src/sync/sync_session.cpp

```diff
@@ -737,28 +739,33 @@ void SyncSession::add_completion_callback(_impl::SyncProgressNotifier::NotifierT
             // notifications from it
             return;
         }
-        auto callbacks = std::move(is_download ? self->m_download_completion_callbacks
-                                               : self->m_upload_completion_callbacks);
+        auto& callbacks = is_download ? self->m_download_completion_callbacks
+                                      : self->m_upload_completion_callbacks;
+        auto callback_node = callbacks.extract(id);
         lock.unlock();
-        for (auto& callback : callbacks)
-            callback(ec);
+        if (callback_node)
+            callback_node.mapped()(ec);
     });
 }
 
 void SyncSession::wait_for_upload_completion(std::function<void(std::error_code)> callback)
 {
     std::unique_lock<std::mutex> lock(m_state_mutex);
-    if (m_upload_completion_callbacks.empty())
-        m_state->wait_for_completion(*this, _impl::SyncProgressNotifier::NotifierType::upload);
-    m_upload_completion_callbacks.push_back(std::move(callback));
+    m_completion_request_counter++; // read in wait_for_completion()
+    m_state->wait_for_completion(*this, _impl::SyncProgressNotifier::NotifierType::upload);
+    m_upload_completion_callbacks.emplace_hint(m_upload_completion_callbacks.end(),
```

Comment:
Do you think it is worth moving this into `add_completion_callback`? I can't really see any reason for the current distribution of work between the functions, but I didn't want to make too much unnecessary change in a bugfix commit.

Reply:
Do we still need separate upload/download completion callback maps with this change? If not then merging them and pushing this into add_completion_callback() makes sense, but without merging them it'd be another awkward conditional there?

Reply:
No, I don't think so. Merging them cleans up the code nicely.